flink/src/main/java/org/apache/iceberg/flink/source/DataIterator.java

```diff
@@ -71,12 +72,13 @@
 
   InputFile getInputFile(FileScanTask task) {
     Preconditions.checkArgument(!task.isDataTask(), "Invalid task type");
-
-    return inputFiles.get(task.file().path().toString());
+    return getInputFile(task.file().path().toString());
   }
 
   InputFile getInputFile(String location) {
-    return inputFiles.get(location);
+    // normalize the path before looking it up in the map
+    Path path = new Path(location);
```

Comment:
I don't think using the Hadoop API directly is a good way to solve the problem. It sounds like we need to fix the keys in the map to match the original location from the input split instead.

Reply:
For @steveloughran's information, the reason the keys in the `inputFiles` map became normalized paths is that the files for the scan tasks go through encryption and decryption:
```
    Stream<EncryptedInputFile> encrypted = keyMetadata.entrySet().stream()
        .map(entry -> EncryptedFiles.encryptedInput(io.newInputFile(entry.getKey()), entry.getValue()));

    // decrypt with the batch call to avoid multiple RPCs to a key server, if possible
    Iterable<InputFile> decryptedFiles = encryptionManager.decrypt(encrypted::iterator);
```
and the keys are the location of the decrypted files. The call to `io.newInputFile` either passes through `HadoopFileIO` or `S3FileIO` (currently the two implementations of `FileIO`), and the `newInputFile` process normalized the path. The normalization through `HadoopFileIO` is exactly what I'm replicating here. I think it will work for an S3 path too.

@rdblue, from the decrypted files, I do not see a natural way to recover the original path string written in the manifest. Instead, can we add a method to the `FileIO` interface to return the normalized path for a path String, and then `HadoopFileIO` and `S3FileIO` will have to implement it?

Reply:
Sounds like we will need to fix the interfaces that are modifying the file location and pass the original.

Reply:
I think that `HadoopInputFile` just needs to be updated so that `fromLocation` preserves the original location rather than using `path.toString()`.

Reply:
@rdblue thank you for the suggestion; I have implemented it and updated the PR description accordingly.

Reply:
According to the javadoc for `InputFile#location`, this is "The fully-qualified location of the input file as a String." The only concern I have is if we ever call `FileIO#newInputFile` or `HadoopInputFile.fromLocation` with a location that is not fully-qualified.